File: week3/chanki/백준 20366. 같이 눈사람 만들래?.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

// 문제 풀이에 필요한 아이디어
// 눈덩이를 크기 순으로 정렬하고 A B C D 가 있다면
// 두 눈덩이 간 크기의 최소 차는 A, D / B, C로 구성될 때다.
public class B20366 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int snowballCount = Integer.parseInt(br.readLine());
        List<Integer> radiusOfSnowman = Arrays.stream(br.readLine().split(" "))
            .map(Integer::parseInt)
            .sorted()
            .collect(Collectors.toList());

        int minimumDiff = Integer.MAX_VALUE;

        outer:
        // 바깥 눈덩이를 가능한 케이스대로 모두 순회
        for (int outerLeft = 0; outerLeft < snowballCount - 3; outerLeft++) {
            for (int outerRight = snowballCount - 1; outerLeft + 2 < outerRight; outerRight--) {
                int innerLeft = outerLeft + 1;
                int innerRight = outerRight - 1;
                // 안쪽 눈덩이는 투포인터로 진행
                while (innerLeft < innerRight) {
                    int innerSnowmanSize = radiusOfSnowman.get(innerLeft) + radiusOfSnowman.get(innerRight);
                    int outerSnowmanSize = radiusOfSnowman.get(outerLeft) + radiusOfSnowman.get(outerRight);
                    int snowmanSizeDiff = outerSnowmanSize - innerSnowmanSize;

                    minimumDiff = Math.abs(minimumDiff) < Math.abs(snowmanSizeDiff)
                        ? minimumDiff
                        : snowmanSizeDiff;

                    // 차이가 0인 경우 더 작을 수는 없으므로 모든 순회를 종료
                    // outer - left > 0 이면 inner가 커져야 하므로, innerLeft++
                    if (snowmanSizeDiff == 0) {
                        break outer;
                    } else if (snowmanSizeDiff > 0) {
                        innerLeft++;
                    } else {
                        innerRight--;
                    }
                }
            }
        }

        System.out.println(Math.abs(minimumDiff));
    }
}

File: week3/chanki/백준 20366. 같이 눈사람 만들래?.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import kotlin.math.abs

// 문제 풀이에 필요한 아이디어
// 눈덩이를 크기 순으로 정렬하고 A B C D 가 있다면
// 두 눈덩이 간 크기의 최소 차는 A, D / B, C로 구성될 때다.
fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val snowballCount = readLine().toInt()
    val radiusOfSnowball = readLine().split(" ").map { it.toInt() }.sorted().toList()

    var minimumDiff = Integer.MAX_VALUE
    // 레이블을 설정하여 차이가 0인 경우 모든 순회 종료
    // 바깥의 두 눈덩이는 실행 가능한 모든 경우를 실행
    outer@ for (outerLeft in 0..<snowballCount - 3) {
        for (outerRight in snowballCount - 1 downTo outerLeft + 3) {
            val outerSnowmanSize = radiusOfSnowball[outerLeft] + radiusOfSnowball[outerRight]
            var innerLeft = outerLeft + 1
            var innerRight = outerRight - 1

            // 안에 있는 눈덩이는 투포인터로 최적화
            while (innerLeft < innerRight) {
                val innerSnowmanSize = radiusOfSnowball[innerLeft] + radiusOfSnowball[innerRight]
                val snowmanSizeDiff = outerSnowmanSize - innerSnowmanSize

                minimumDiff = minimumDiff.coerceAtMost(abs(snowmanSizeDiff))

                when {
                    snowmanSizeDiff == 0 -> break@outer
                    snowmanSizeDiff > 0 -> innerLeft++
                    else -> innerRight--
                }
            }
        }
    }

    println(minimumDiff)
}
